### PR TITLE
#10100: Add support for paged update_cache and fill_cache

### PR DESCRIPTION
**Pull request by cglagovichTT**
### Ticket
#10100 
I'm reopening this PR because it broke build and was reverted.

### Problem description
Please see issue for detailed description.

To support paged attention, we need update_cache and fill_cache to take a page_table tensor for indirect indexing into the paged KV cache.

### What's changed

This PR adds support in paged_update_cache and paged_fill_cache to take a page_table tensor such that these ops support paged KV caches.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/10461362650/job/28969517045
- [ ] Model regression CI testing passes (if applicable) (N/A)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by tt-rkim** [comment]
Gonna defer to you on this review lol @ayerofieiev-tt

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        return cache_tensor; // Updated cache tensor in-place
+    }
+};
+
+}  // namespace operations::experimental::paged_cache
+
+namespace experimental::paged_cache {
+
+constexpr auto paged_update_cache = ttnn::register_operation_with_auto_launch_op<
+    "ttnn::experimental::paged_cache::paged_update_cache",
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:40]:
```suggestion
    "ttnn::experimental::paged_update_cache",
```

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    ttnn::operations::experimental::paged_cache::PagedUpdateCacheOperation>();
+
+constexpr auto paged_fill_cache = ttnn::register_operation_with_auto_launch_op<
+    "ttnn::experimental::paged_cache::paged_fill_cache",
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:44]:
```suggestion
    "ttnn::experimental::paged_fill_cache",
```

the agreement is to make things accessible in Python as `ttnn.experimental.<op>`, similarly to how categories non-experimental ops are in `ttnn.<op>`

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        auto kernel_config_val = init_device_compute_kernel_config(input_tensor.device()->arch(), compute_kernel_config);
+        operation::run(PagedUpdateCacheDeviceOperation{batch_idx, {}, 0, PagedUpdateCacheOpType::FILL, kernel_config_val}, {cache_tensor, input_tensor, page_table}, {std::nullopt, std::nullopt});
+
+        return cache_tensor; // Updated cache tensor in-place
+    }
+};
+
+}  // namespace operations::experimental::paged_cache
+
+namespace experimental::paged_cache {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:37]:
```suggestion
namespace experimental{
```

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+#include "device/paged_cache_operation.hpp" // TODO: not right!
+#include "ttnn/run_operation.hpp"
+#include "ttnn/deprecated/tt_dnn/op_library/compute_kernel_config.hpp"
+#include "ttnn/core.hpp"
+
+namespace ttnn {
+namespace operations::experimental::paged_cache {
+
+struct PagedUpdateCacheOperation {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:15]: Can we split these into two?

* ttnn
  * operations
    * experimental
      * paged_cache
        * common (if needed)
        * paged_fill_cache
        * paged_update_cache

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:15]: They have different kernels, programs, device operations.

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    operation::ProgramWithCallbacks create_program(
+        const std::vector<Tensor>& input_tensors,
+        const std::vector<std::optional<const Tensor>>& optional_input_tensors,
+        std::vector<Tensor>& output_tensors
+    ) const;
+
+    static constexpr auto attribute_names =
+        std::forward_as_tuple("batch_idx", "update_idxs", "batch_offset", "op_type", "compute_kernel_config");
+
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.hpp:48]: afaik this is not needed anymore
Check with @arakhmati

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp (hunk 4 above)
Comment (cglagovichTT) [reply ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:15]: They share the same op device struct, just like `update_cache` and `fill_cache`.

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:15]: Just a suggestion, if you think current structure is more proper, I am good

Comment (cglagovichTT) [reply ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:15]: I appreciate the thorough look you're giving the code :) I think the current structure is alright - there's just enough shared code between the two ops that it makes sense to me to have everything but the program factory be shared.

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.hpp (hunk 5 above)
Comment (sraizada-tt) [reply ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_cache_operation.hpp:48]: Akhmed: "after updating to C++20, we finally could start using https://github.com/qlibs/reflect. It lets us to iterate over the member variables of structs. So, we were able to delete a lot of the uses of attribute_names.
However, the library only works for simple structs without custom constructors, so we couldn't get rid of all the places where we use attribute_names yet. And therefore you might need it in some cases"

**Comment by sraizada-tt** [comment]
@ayerofieiev-tt Can I get a re-review please?

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache_pybind.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+#include "ttnn/operations/experimental/paged_cache/paged_cache.hpp"
+#include "ttnn/deprecated/tt_dnn/op_library/compute_kernel_config.hpp"
+
+
+namespace ttnn::operations::experimental::paged_cache::detail {
+
+namespace py = pybind11;
+
+void bind_experimental_paged_cache_operations(py::module& module) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache_pybind.hpp:20]: please split into .hpp and .cpp 
follow example from ccp/all_gather

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_fill_cache_program_factory.hpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+#include "paged_cache_operation.hpp"
+#include "ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/work_split.hpp"
+
+namespace ttnn::operations::experimental::paged_cache::detail {
+
+using namespace tt::constants;
+using namespace tt;
+
+
+operation::ProgramWithCallbacks paged_fill_cache_multi_core(const Tensor& cache_tensor, const Tensor &input_tensor, const Tensor &page_table_tensor, const uint32_t batch_idx) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_fill_cache_program_factory.hpp:18]: split to hpp and cpp please

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_update_cache_program_factory.hpp [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+#include "paged_cache_operation.hpp"
+#include "ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/work_split.hpp"
+
+
+namespace ttnn::operations::experimental::paged_cache::detail {
+
+using namespace tt::constants;
+using namespace tt;
+
+bool enable_fp32_dest(const tt_metal::Device * device, const DeviceComputeKernelConfig& compute_kernel_config, const tt::DataFormat& input_cb_data_format) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/device/paged_update_cache_program_factory.hpp:18]: split to hpp and cpp please

File: tests/ttnn/unit_tests/operations/test_paged_update_cache.py
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_paged_update_cache.py:1]: Please replace all usage of tt_lib/ttl with ttnn here. Thank you!

File: ttnn/cpp/ttnn/operations/experimental/experimental_pybind.hpp [hunk 9]
```diff
@@ -5,8 +5,10 @@
 #pragma once
 
 #include "pybind11/pybind_fwd.hpp"
-namespace ttnn::operations::experimental {
+#include <pybind11/pybind11.h>
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/experimental_pybind.hpp:8]: This file should only have fwd to include

File: ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp [hunk 10]
```diff
@@ -0,0 +1,40 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "device/paged_cache_operation.hpp" // TODO: not right!
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/paged_cache/paged_cache.hpp:7]: You should avoid device operation include in the operation header. Think of it as a detail of implementation. 
if something from it is needed here it is best to have that extracted into <op>_types.hpp or <op>_common.hpp

**Review by eyonland — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**